Conch/lib/Conch/Orc/Lifecycle.pm

```diff
@@ -0,0 +1,472 @@
+	id
+	created
+	updated
+	name
+	version
+	locked
+	role_id
+	plan
+]] }
+
+
+
+=head1 METHODS
+
+=head2 from_id
+
+Load a Lifecycle by its UUID
+
+=cut
+
+sub from_id ($class, $id) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null and l.id = ?;
+		|, $id)->hash;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->from_id: $_");
+		return undef;
+	};
+
+	return undef unless $ret;
+	return $class->new(_fixup_timestamptzs($ret)->%*);
+}
+
+
+=head2 from_name_latest
+
+Load the latest version of a Lifecycle for a given name
+
+=cut
+
+sub from_name_latest ($class, $name) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null and l.id = ?;
+				order by l.version
+				limit 1
+		|, $name)->hash;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->from_name_latest: $_");
+		return undef;
+	};
+
+	return undef unless $ret;
+	return $class->new(_fixup_timestamptzs($ret)->%*);
+}
+
+
+=head2 many_from_name
+
+Load all Lifecycles that match a given name
+
+=cut
+
+sub many_from_name ($class, $name) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null and l.name = ?;
+		|, $name)->hashes;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->many_from_name: $_");
+		return undef;
+	};
+
+	return [] unless $ret and $ret->@*;
+
+	my @many = map {
+		$class->new(_fixup_timestamptzs($_)->%*);
+	} $ret->@*;
+
+	return \@many;
+}
+
+
+
+=head2 all
+
+Returns an arrayref containing all active Lifecycles
+
+=cut
+
+sub all ($class) {
+	my $ret;
+	try {
+		$ret = Conch::Pg->new->db->query(qq|
+			select l.*, array(
+				select wlp.workflow_id
+				from workflow_lifecycle_plan wlp
+				where wlp.lifecycle_id = l.id
+				order by wlp.plan_order
+			) as plan
+			from workflow_lifecycle l
+			where l.deactivated is null;
+		|)->hashes;
+	} catch {
+		Mojo::Exception->throw(__PACKAGE__."->all: $_");
+		return undef;
+	};
+	
+	return [] unless $ret and $ret->@*;
+
+	my @many = map {
+		$class->new(_fixup_timestamptzs($_)->%*);
+	} $ret->@*;
+
+	return \@many;
+}
+
+
+=head2 save
+
+Save or update a Lifecycle in the database. The plan is B<not> saved this way.
+
+Returns C<$self>
+
+=cut
+
+sub save ($self) {
+	my $db = Conch::Pg->new()->db;
+	$self->_set_updated(Conch::Time->now);
+
+	my %fields = $self->%*;
+	delete $fields{id};
+	delete $fields{plan};
```

Comment:
Should deactivated be allowed to be set? Created and updated should be managed by Postgres.